### PR TITLE
Decode RAW_DATA and LOG_DATA push notifications into structured events

### PR DESCRIPTION
**Pull request by f4fez**
## Summary

`PacketType::RawData` (0x84) was previously received and silently dropped (`PacketType::RawData => {}`). This PR:

- Adds `RouteType`/`PayloadType` enums (`packets.rs`) decoding the 1-byte mesh packet header (route type, payload type, payload version), matching the existing `BinaryReqType`/`ControlType`/`AnonReqType` style.
- Adds `MeshPacketHeader` and `RawAdvertisement` (`events.rs`), used to decode **both** `RawData` (`RawPacketData`) and `LogData`: SNR/RSSI, the header, the transport code (if present), the hop path, and — best-effort — the advertiser's identity when `payload_type` is `Advert`. Other payload types (direct/channel messages) stay as opaque bytes in `payload`, since decoding them needs per-channel secrets this reader has no access to.
- Adds `parse_mesh_packet_header()` and `parse_raw_advertisement()` to `parsing.rs`.
- Adds `examples/rf_packet_monitor.rs` demonstrating a live monitor over serial.
- Trims the `tokio` dependency from `features = ["full", ...]` down to `["sync", "time", "macros", "rt"]`, with `io-util`/`net` only pulled in by the transports that need them (`serial`/`tcp`), plus `rt-multi-thread`/`signal` moved to `[dev-dependencies]` (examples/doctests only). Downstream consumers currently get tokio's full feature set unconditionally via feature unification; this removes that.

## Important: RAW_DATA vs LOG_DATA

I initially assumed `RawData` was a general "hear every packet" monitor. Verifying against the firmware itself (`meshcore-dev/MeshCore`, `Dispatcher.cpp` / `examples/companion_radio/MyMesh.cpp`) showed that's wrong:

- **`LOG_DATA`** (`PUSH_CODE_LOG_RX_DATA`) is pushed unconditionally by `Dispatcher::checkRecv()` for *every* packet the radio receives, regardless of payload type or routing — no activation needed. This is the general RF monitor, and what `examples/rf_packet_monitor.rs` subscribes to.
- **`RAW_DATA`** is much narrower: `onRawDataRecv()` only fires for directly-routed, not-yet-seen `PAYLOAD_TYPE_RAW_CUSTOM` packets (sent by another app via the companion `SEND_RAW_DATA` command). Regular mesh traffic never triggers it. `RawPacketData`'s doc comment calls this out explicitly to avoid the same mix-up for future readers.

The wire format (SNR/RSSI[/reserved for RAW_DATA], then header+path+payload) and the packet header bit layout (route_type: bits 0-1, payload_type: bits 2-5, payload_version: bits 6-7, then optional 4-byte transport code, then a path-length/hash-size byte and path) were verified against the reference implementation in `meshcore_py` (`reader.py` `PacketType.RAW_DATA`/`LOG_DATA` and `meshcore_parser.py` `parsePacketPayload`), which this crate is already a port of.

## Test plan

- [x] 44 new unit tests covering the new enums, the two new parsing functions, and the reader wiring for both `RAW_DATA` and `LOG_DATA` (including short-payload edge cases) — 268 tests total, all passing
- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --tests --no-deps --all-features --all-targets`
- [x] `cargo build --release` and `cargo publish --dry-run --allow-dirty --no-verify`
- [x] `cargo check --no-default-features` and each of `--features {serial,tcp,ble}` individually, to validate the tokio feature trim
- [x] `cargo tree -e features -i tokio` confirms `fs`/`io-std`/`process`/`parking_lot`/`test-util` are no longer pulled in
- [x] `examples/rf_packet_monitor.rs` run live against a real MeshCore node (BLE and serial), confirmed `LogData` events with decoded header/path/advertiser arrive as expected

Opened as a draft since this bundles two logically-separate changes (the RAW_DATA/LOG_DATA decoding, and the tokio feature trim) — happy to split into two PRs if preferred.

### CODE REVIEW

**Comment by andrewdavidmackenzie** [comment]
Thanks for the contribution. Reviewing in detail now.

BTW: It would be nice to hear your use case for wanting this functionality (I agree it was missing, no argument there...).

It looks like you're mainly using Serial?

Also "nice" (but not a must) to hear how you are using meshcore-rs.

Nice catch on the dependency trim! I usually try to avoid that. I use all-features to test all feature combinations build, and (sporadically) I manually use udeps to find unused dependencies. I guess I'd need to run udeps on all feature combinations to find examples like this automatically.

I think it would be good to have that in a separate PR, apart from code changes, for git log cleanliness. If you find a good way to avoid such cases as that in the future as part of CI (that doesn't add too much time), I'd be interested and would accept in the same PR.

Now that I have enabled the workflows to run, it shows 0% test coverage of the example added.
Just 2 lines missing on parsing.rs changes, but maybe two that are important to test?

CodeRabbit has chipped in with it's review, with some nits. I usually find them pretty good, even if nits, so take a look and either push changes or comment in the review for me.

I'll add any of my own review comments I have, now.

Thanks again

File: examples/rf_packet_monitor.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+//! Example showing how to monitor all RF packets received by the node
```
Comment (andrewdavidmackenzie) [inline examples/rf_packet_monitor.rs:1]: Could maybe make explicit it uses serial.

I think I'd also rename to make that obvious from a glance at the examples folder?

"rf_packet_monitor_serial.rs" maybe?

A nice enhancement for people like myself who don't have any serial device (this applies to some other examples too...) would be to ask the user (or take an option) to specify how to connect, serial, wifi, bt - etc.

That would be nice to factor out as a util module all the examples could use....if you don't want to tackle that I understand and I'd create an issue to suggest it.

File: examples/rf_packet_monitor.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                                || header.route_type == RouteType::TransportDirect
+                            {
+                                if let Some(code) = header.transport_code {
+                                    println!("  Transport code: {:02x?}", code);
+                                }
+                            }
+
+                            if header.payload_type == PayloadType::Advert {
+                                if let Some(adv) = log.advertisement {
+                                    println!("  Advertiser: {:02x?}", &adv.public_key[..6]);
```
Comment (andrewdavidmackenzie) [inline examples/rf_packet_monitor.rs]: I expect coderabbit, claude, maybe even clippy to suggest some changes to reduce the nested ifs....

File: examples/rf_packet_monitor.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        .nth(1)
+        .unwrap_or_else(|| "/dev/ttyUSB0".to_string());
+
+    println!("Connecting to MeshCore device on {port}...");
+    let meshcore = MeshCore::serial(&port, 115200).await?;
+
+    let self_info = meshcore.commands().lock().await.send_appstart().await?;
+    println!("Connected to device: {}", self_info.name);
+
+    let _sub = meshcore
```
Comment (andrewdavidmackenzie) [inline examples/rf_packet_monitor.rs:52]: It would make main() more readable if this was moved to a function. Not a big deal for an example (many examples in projects seem to be one big long function...) but would make the "subscribe and wait for control-c" "loop" very visible.

File: examples/rf_packet_monitor.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    tracing_subscriber::fmt()
+        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+        .init();
+
+    let port = env::args()
+        .nth(1)
+        .unwrap_or_else(|| "/dev/ttyUSB0".to_string());
+
+    println!("Connecting to MeshCore device on {port}...");
+    let meshcore = MeshCore::serial(&port, 115200).await?;
```
Comment (andrewdavidmackenzie) [inline examples/rf_packet_monitor.rs]: This would be the chunk subject to being factored out and accepting other connection methods as args.

File: src/events.rs [hunk 5]
```diff
@@ -1447,6 +1533,8 @@ mod tests {
         let log_data = LogData {
             snr: 12.25,
             rssi: -75,
+            header: None,
```
Comment (andrewdavidmackenzie) [inline src/events.rs]: When I see many of these as None, None - it makes me think it would be good to provide a Default implementation for LogData, and then these other fields could just be elipsized on creation, shorting the test code when not setting those fields to anything specific.

File: src/events.rs [hunk 6]
```diff
@@ -1495,6 +1589,8 @@ mod tests {
         let log_data = LogData {
```
Comment (andrewdavidmackenzie) [inline src/events.rs]: e.g. this could be just LogData::default()

File: src/packets.rs [hunk 7]
```diff
@@ -187,6 +187,87 @@ impl From<u8> for ControlType {
...
+    Flood = 1,
+    /// Directly-routed packet along an explicit path
+    Direct = 2,
+    /// Directly-routed packet that also carries a transport code
+    TransportDirect = 3,
+}
+
+impl From<u8> for RouteType {
+    fn from(value: u8) -> Self {
+        match value & 0x03 {
```
Comment (andrewdavidmackenzie) [inline src/packets.rs:208]: Strictly speaking I don't think you need the "& 0x03" since you have _, but I get it.

File: src/parsing.rs [hunk 8]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+///
+/// Returns the decoded header together with the remaining, unparsed inner
+/// packet payload, or `None` if `data` is too short to contain a header and
+/// a path byte.
+pub fn parse_mesh_packet_header(data: &[u8]) -> Option<(MeshPacketHeader, &[u8])> {
+    if data.is_empty() {
+        return None;
+    }
+
+    let header_byte = data[0];
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: This is my "jonesy" tool being pedantic. I guess data could be 0 length?
it probably recommends .get(0) and handling the error to return None.

File: src/parsing.rs [hunk 9]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
 }
 
+/// Parse the MeshCore over-the-air packet header (route type, payload type
+/// and path) from the start of a buffer, as embedded in RAW_DATA/LOG_DATA
+/// captures.
+///
+/// Returns the decoded header together with the remaining, unparsed inner
+/// packet payload, or `None` if `data` is too short to contain a header and
+/// a path byte.
+pub fn parse_mesh_packet_header(data: &[u8]) -> Option<(MeshPacketHeader, &[u8])> {
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs:618]: Is there a comment describing the bit layout of the header byte somewhere, to help understand the bitshifts and masking below?

File: src/parsing.rs [hunk 10]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+        RouteType::TransportFlood | RouteType::TransportDirect
+    ) {
+        let code: [u8; 4] = read_bytes(data, offset).ok()?;
+        offset += 4;
+        Some(code)
+    } else {
+        None
+    };
+
+    if offset >= data.len() {
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: you could use .get() here to simplify the code and placate jonesy :-)

Maybe use it with a literal '5' to avoid the integer arithmetic above?

code_offset = 1
path_offset = 5

or declare CONSTS alongside the struct definition even better?

File: src/parsing.rs [hunk 11]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+    };
+
+    Some((header, &data[offset..]))
+}
+
+/// Parse a raw ADVERT payload (public key, timestamp, signature, flags and
+/// optional location/name), as carried by a [`PayloadType::Advert`] packet.
+pub fn parse_raw_advertisement(data: &[u8]) -> Option<RawAdvertisement> {
+    // public_key(32) + timestamp(4) + signature(64) + flags(1)
+    const MIN_LEN: usize = 32 + 4 + 64 + 1;
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: define constants for the 32, 4, 64, 1 to help understand the packet makeup?

File: src/parsing.rs [hunk 12]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+    if data.len() < MIN_LEN {
+        return None;
+    }
+
+    let public_key: [u8; 32] = read_bytes(data, 0).ok()?;
+    let timestamp = read_u32_le(data, 32).ok()?;
+    let signature: [u8; 64] = read_bytes(data, 36).ok()?;
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: and use the consts (or sums of them) here

File: src/parsing.rs [hunk 13]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+    let flags = data[100];
+    let adv_type = flags & 0x0F;
+
+    let mut offset = MIN_LEN;
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs:704]: I assume this is that flags is the last 1 byte, hence use of MIN_LEN, but not very descriptive? declare consts for each of the offsets based on sums of the length consts needed to get to it?

File: src/parsing.rs [hunk 14]
```diff
@@ -591,6 +592,117 @@ pub fn parse_mma(data: &[u8]) -> Vec<MmaEntry> {
...
+
+    let (lat, lon) = if flags & 0x10 != 0 {
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: I'd like consts defined for masks, with a comment describing each briefly (if not obvious from the name)

File: src/parsing.rs [hunk 15]
```diff
@@ -1236,4 +1348,130 @@ mod tests {
...
+        assert_eq!(header.payload_version, 0);
+        assert!(header.transport_code.is_none());
+        assert_eq!(header.path_len, 2);
+        assert_eq!(header.path_hash_size, 1);
+        assert_eq!(header.path, vec![0xAA, 0xBB]);
+        assert_eq!(remaining, &[0xCC, 0xDD]);
+    }
+
+    #[test]
+    fn test_parse_mesh_packet_header_with_transport_code() {
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs:1410]: A lot of literals in the tests too. COuld some of these be consts defined alognside the structs to help understand each one (you have tried to make up for that with comments)...

File: src/reader.rs [hunk 16]
```diff
@@ -651,28 +651,86 @@ impl MessageReader {
             PacketType::PathDiscovery => {}
             PacketType::SetFloodScope => {}
             PacketType::SendControlData => {}
-            PacketType::RawData => {}
-            PacketType::LogData => {
-                // LOG_DATA format:
+            PacketType::RawData => {
```
Comment (andrewdavidmackenzie) [inline src/reader.rs:732]: You've followed my (bad) example above, with the code inline.... I think best we factor each one out to a function with a brief comment header describing it - it will make this overall function much shorter and easier to understand. 

It could also facilitate unit tests of this logic?

File: src/reader.rs
Comment (andrewdavidmackenzie) [inline src/reader.rs:688]: Yuk, I will create an issue to improve this code and try to lead by (a better) example

Comment (andrewdavidmackenzie) [reply src/reader.rs:688]: Issue here https://github.com/andrewdavidmackenzie/meshcore-rs/issues/60

**Review by andrewdavidmackenzie — CHANGES_REQUESTED** [review]
I see quite a few things to improve that reflect more on the code style you followed, that I should improve. 

There are some other points I think worth considering too.

And in comments I suggested splitting out the dependency improvements to another PR:

thanks!

**Comment by andrewdavidmackenzie** [comment]
## Review generated by Claude Code

### Overall Assessment

This is a high-quality PR with thorough research, good test coverage, and clear documentation. The firmware analysis in the PR description is accurate. The LOG_DATA decoding is correct and complete. The tokio feature trim is clean.

The one real issue is the RAW_DATA handler applying `parse_mesh_packet_header()` to what is actually opaque application-level data, not a full mesh packet. The PR author clearly understands the distinction (the doc comments describe it correctly), but the code doesn't reflect it — it parses both identically. The `RawPacketData` struct should either drop the `header`/`advertisement` fields or the handler should skip the parse and treat the payload as opaque bytes.

---

### Detailed Findings

#### Protocol Correctness (LOG_DATA) — Correct

The LOG_DATA handler is faithfully ported from the Python reference (`meshcore_py`). Verified against:

- **Firmware** (`Dispatcher.cpp`): `logRxRaw()` is called unconditionally in `checkRecv()` for every received packet, BEFORE parsing. The companion radio (`MyMesh.cpp`) serializes it as: `[push_code, SNR*4, RSSI, raw_bytes...]` — no reserved byte.
- **Python reference** (`reader.py`): Reads SNR (byte 0, signed/4.0), RSSI (byte 1, signed), then passes remaining bytes to `parsePacketPayload()`.
- **Rust PR**: Matches exactly. Bytes 0-1 are SNR/RSSI, bytes 2+ are the full over-the-air mesh packet, which is correctly passed to `parse_mesh_packet_header()`.

#### Protocol Correctness (RAW_DATA) — Semantic Issue

There is a semantic error in the RAW_DATA handler. The PR applies `parse_mesh_packet_header()` to RAW_DATA payloads identically to LOG_DATA, but the wire contents are fundamentally different:

- **LOG_DATA** sends the full raw over-the-air bytes (mesh header byte + transport codes + path + inner payload) — `parse_mesh_packet_header()` is appropriate.
- **RAW_DATA** sends `packet->payload` — the **already-stripped inner payload** after the firmware has parsed and removed the mesh header byte, transport codes, and path. For `PAYLOAD_TYPE_RAW_CUSTOM`, this is opaque application-level data.

The Python reference confirms this distinction: it does **not** call `parsePacketPayload()` on RAW_DATA — it stores the payload as opaque hex. Only LOG_DATA gets structured parsing.

In practice, since `parse_mesh_packet_header()` returns `None` on malformed data, this won't crash. But:
- `header` will contain **garbage** (interpreting application bytes as a mesh header) or `None`
- `advertisement` will almost always be `None`
- The `payload` field will contain a **subset** of the actual payload (bytes remaining after the "header" was consumed)

**Recommendation**: RAW_DATA should skip the `parse_mesh_packet_header()` call. The `RawPacketData` struct should just have `snr`, `rssi`, and `payload: Vec<u8>` (opaque bytes), without `header` or `advertisement` fields. Alternatively, if the struct is kept as-is for forward compatibility, document that `header`/`advertisement` are always `None` and don't attempt the parse.

Firmware evidence (`MyMesh.cpp`, `onRawDataRecv`):
```cpp
memcpy(&out_frame[i], packet->payload, packet->payload_len);
```
This copies `packet->payload` — the inner payload AFTER `Dispatcher::tryParsePacket()` has already stripped the header byte, transport codes, and path into separate `Packet` fields.

#### Header Bit Layout — Correct

Verified against `Packet.h`:

| Field | Firmware | PR | Status |
|---|---|---|---|
| route_type | `header & 0x03` (bits 0-1) | `RouteType::from(header_byte)` masks `& 0x03` | Match |
| payload_type | `(header >> 2) & 0x0F` (bits 2-5) | `PayloadType::from(header_byte >> 2)` masks `& 0x0F` | Match |
| payload_version | `(header >> 6) & 0x03` (bits 6-7) | `(header_byte & 0xc0) >> 6` | Match |
| path_hash_size | `(path_len >> 6) + 1` | `((path_byte & 0xC0) >> 6) + 1` | Match |
| path_hash_count | `path_len & 63` | `path_byte & 0x3F` | Match |

Route type and payload type enum values all match `Packet.h` constants exactly.

#### PayloadType Enum — Minor Gap

`PAYLOAD_TYPE_RAW_CUSTOM` is defined as `0x0F` in the firmware but maps to `PayloadType::Unknown` in the PR (values 12-15 all become `Unknown`). Since the PR only encounters RAW_CUSTOM implicitly via the RAW_DATA push (which doesn't carry the header byte anyway), this is not a runtime problem. But for LOG_DATA, if you overhear a RAW_CUSTOM packet on the mesh, its payload type would show as `Unknown` rather than a descriptive variant. Adding a `RawCustom = 0x0F` variant (and keeping 12-14 as `Unknown`) would be more complete.

#### Advertisement Parsing — Correct

The `parse_raw_advertisement()` function matches the Python `parsePacketPayload()` for `payload_type == 0x04`:
- `pubkey(32) + timestamp(4) + signature(64) + flags(1)` = 101 bytes minimum
- Flag bits: `0x0F` = adv_type, `0x10` = location (8 bytes), `0x20` = feature1 (2 bytes, skipped), `0x40` = feature2 (2 bytes, skipped), `0x80` = name (remaining bytes)
- The truncated-location regression test (commit 3) is a good addition.

#### Tokio Feature Trim — Correct and Beneficial

- Core: `default-features = false, features = ["sync", "time", "macros", "rt"]` — genuinely the minimum needed.
- `io-util` gated behind `serial` and `tcp`; `net` gated behind `tcp`.
- `rt-multi-thread` + `signal` moved to dev-dependencies.
- Removes `parking_lot` from the dependency tree entirely.

Clean, well-reasoned change.

#### Code Quality

- Follows existing patterns (`RouteType`/`PayloadType` mirror `BinaryReqType`/`ControlType`/`AnonReqType` style).
- Thorough doc comments with cross-references between `RawPacketData`/`LogData`.
- `parse_mesh_packet_header()` and `parse_raw_advertisement()` use `Option<T>` with `?` propagation — clean and consistent.

#### Test Coverage

44 new tests covering enums, both parsing functions (including edge cases: empty, truncated, missing transport code, truncated location), and reader integration for both RAW_DATA and LOG_DATA. Well-constructed with realistic wire-format byte sequences.

---

### Issues Summary

| Issue | Severity | Description |
|---|---|---|
| RAW_DATA `parse_mesh_packet_header()` | **Medium** | Applying mesh header parsing to RAW_DATA payload is semantically wrong — firmware sends already-stripped inner payload, not raw over-the-air bytes. Python reference does NOT parse RAW_DATA payloads. |
| Missing `RawCustom` PayloadType variant | **Low** | `PAYLOAD_TYPE_RAW_CUSTOM` (0x0F) maps to `Unknown` instead of a named variant. |

**Comment by andrewdavidmackenzie** [comment]
I didn't pick-up on this one that Claude flags:

"The one real issue is the RAW_DATA handler applying parse_mesh_packet_header() to what is actually opaque application-level data, not a full mesh packet. The PR author clearly understands the distinction (the doc comments describe it correctly), but the code doesn't reflect it — it parses both identically. The RawPacketData struct should either drop the header/advertisement fields or the handler should skip the parse and treat the payload as opaque bytes."

can you comment on that?

**Comment by f4fez** [comment]
Hi,
You are welcome 😉. Thank you for providing this crate.
I will separate the pr and refine them.
To explain the usage, I just started an application to monitor a cluster of repeater. I want to log raw packets in a TUI. And maybe in the futur send them to a mqtt broker to replace the python observer.
I send you the Link later. I'm answering you from my phone at work and it's not confortable 😕

File: examples/rf_packet_monitor.rs (hunk 1 above)
Comment (f4fez) [reply examples/rf_packet_monitor.rs:1]: It was to keep it simple for the example. I'm agree a simple toolbox module can work and do not add complexity to understand. Setting the connection as simple parameters like "cargo --example rf_packet_monitor ble My_mesh_controller" for start ?

Comment (andrewdavidmackenzie) [reply examples/rf_packet_monitor.rs:1]: yes. It might need a "--" in the middle to pass the args to the example, but yes. A small addition to the docs or README to explain the options would be great!

--example {example_name} -- [serial | bt | tcp] {tty for serial | bluetooth device name | IP?}

something like that.

That would be a nice little addition, thanks!

File: examples/rf_packet_monitor.rs (hunk 4 above)
Comment (f4fez) [reply examples/rf_packet_monitor.rs]: @andrewdavidmackenzie Can we go for something like cargo --example rf_packet_monitor ble My_mesh_controller" ?

File: src/packets.rs (hunk 7 above)
Comment (f4fez) [reply src/packets.rs:208]: I don't have your point. so I ask Claude:

> The mask is actually load-bearing here, not just defensive. RouteType::from is called with the full header byte (RouteType::from(header_byte) in parsing.rs), where route_type only occupies bits 0-1 — payload_type and payload_version live in bits 2-7. Without & 0x03, a byte like 0b00000100 (payload_type=1, route bits=00 → should be TransportFlood) would fall into the _ arm and silently decode as TransportDirect. The _ gives exhaustiveness, but only the mask guarantees we're matching the right bits. Happy to add a comment noting this if it'd help future readers.

Comment (andrewdavidmackenzie) [reply src/packets.rs:208]: Thinking about it more, I think it depends what other bits might be mixed in with these two lower bytes we're interested in, in the upper 6 bytes.

If they are maybe non-zero, then you are right, we should leave the mask. Sorry for that.

File: src/events.rs (hunk 5 above)
Comment (f4fez) [reply src/events.rs]: Fix. Use default.
I let you close the conversation if ok for you

File: src/events.rs (hunk 6 above)
Comment (f4fez) [reply src/events.rs]: Fix

File: src/parsing.rs (hunk 8 above)
Comment (f4fez) [reply src/parsing.rs]: Hum, don't know this tool. Let me take a look and reply later ;)

File: src/parsing.rs (hunk 11 above)
Comment (f4fez) [reply src/parsing.rs]: Added const to this function.

I let you close the conversation if ok for you ?

File: src/parsing.rs (hunk 14 above)
Comment (f4fez) [reply src/parsing.rs]: Same as previous conversion about constants, closing the topic

File: src/parsing.rs (hunk 13 above)
Comment (f4fez) [reply src/parsing.rs:704]: Added constants to this function also.

I let you close the conversation if ok for you ?

File: src/reader.rs
Comment (f4fez) [reply src/reader.rs:688]: Ok. Can we flag resolve this conversation since it will be done as an other issue ?

File: src/parsing.rs (hunk 9 above)
Comment (f4fez) [reply src/parsing.rs:618]: Comment added. 

I let you close the conversation if ok for you ?

File: examples/rf_packet_monitor.rs (hunk 1 above)
Comment (f4fez) [reply examples/rf_packet_monitor.rs:1]: Added this to easy run examples. No new dependency to a command line parser and simple code to serve the example. In you need something more advanced in the futur, you can easily migrate to cli parsing crate

File: src/parsing.rs (hunk 15 above)
Comment (f4fez) [reply src/parsing.rs:1410]: Added some constants

File: src/reader.rs (hunk 16 above)
Comment (f4fez) [reply src/reader.rs:732]: Functions moved out

**Comment by f4fez** [comment]
> I didn't pick-up on this one that Claude flags:
> 
> "The one real issue is the RAW_DATA handler applying parse_mesh_packet_header() to what is actually opaque application-level data, not a full mesh packet. The PR author clearly understands the distinction (the doc comments describe it correctly), but the code doesn't reflect it — it parses both identically. The RawPacketData struct should either drop the header/advertisement fields or the handler should skip the parse and treat the payload as opaque bytes."
> 
> can you comment on that?

I confirm the bug. I push the fix with the others but one commit per change to simplify the review

File: src/parsing.rs (hunk 8 above)
Comment (andrewdavidmackenzie) [reply src/parsing.rs]: Its a pet project to try and detect all possible sources of panics in code, flag them (comment line tool, GH Action, IDE language server plugin). Then you either change the code to avoid/handle the potential panic, or add a comment or config line to allow it.

This example is: any array indexing can panic. You can use data.get(0) and handle the possible Option/Error.

File: src/parsing.rs (hunk 10 above)
Comment (andrewdavidmackenzie) [reply src/parsing.rs]: Same as above. A pain I know, but this is the project where I really tried to use jonesy "for real" and to avoid as many _potential_ panics as possible and make meshcore-rs as reliable as possible.

(should try fuzz testing it sometime to see if it is....)

**Comment by andrewdavidmackenzie** [comment]
Great progress!

A pain I know, when you have a working PR, but being the first one there are a few "conventions" to get over....

I think it will be great when the few remaining points are addressed, and then I will merge.

File: src/reader.rs [hunk 17]
```diff
@@ -654,42 +654,31 @@ impl MessageReader {
...
+                // See meshcore_py reader.py PacketType.RAW_DATA for the
+                // reference implementation this is ported from.
                 if payload.len() >= 2 {
                     let snr_byte = payload[0] as i8;
                     let snr = snr_byte as f32 / 4.0;
                     let rssi = payload[1] as i8 as i16;
 
-                    let packet = if payload.len() > 3 {
-                        &payload[3..]
+                    let inner_payload = if payload.len() > 3 {
```
Comment (andrewdavidmackenzie) [inline src/reader.rs]: Still a few magic numbers in here. Did you add consts for the offsets with descriptive names?

File: src/parsing.rs [hunk 18]
```diff
@@ -631,10 +627,7 @@ pub fn parse_mesh_packet_header(data: &[u8]) -> Option<(MeshPacketHeader, &[u8])
         None
     };
 
-    if offset >= data.len() {
-        return None;
-    }
-    let path_byte = data[offset];
+    let path_byte = *data.get(offset)?;
     offset += 1;
```
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: rather than doing arithemtic on the offset, I would just use a set of consts, each one with the offset for the part of the packet you are looking for, with the consts defined alongside the structs.

File: src/parsing.rs
Comment (andrewdavidmackenzie) [inline src/parsing.rs]: I'd use consts for all parts of the payload, rather than adding to offset.

File: src/parsing.rs (hunk 8 above)
Comment (f4fez) [reply src/parsing.rs]: Fixed in 7d0bad1 — replaced the raw index with `data.first()?` in `parse_mesh_packet_header` (and `.get(FLAGS_OFFSET)?` in `parse_raw_advertisement`), so a 0-length/truncated buffer now returns `None` instead of panicking.

File: src/parsing.rs (hunk 10 above)
Comment (f4fez) [reply src/parsing.rs]: Fixed in 7d0bad1 — added `HEADER_BYTE_LEN`/`TRANSPORT_CODE_LEN`/`PATH_BYTE_LEN` consts next to `parse_mesh_packet_header` and switched all offset arithmetic to `checked_add`, so an overflow/out-of-bounds read returns `None` instead of panicking.

File: src/reader.rs (hunk 17 above)
Comment (f4fez) [reply src/reader.rs]: Fixed in 7d0bad1 — added `SNR_RSSI_LEN`/`RAW_DATA_RESERVED_LEN` consts and switched `parse_raw_data`/`parse_log_data` to `.first()?`/`.get()` instead of raw indexing and magic-number slicing.

File: src/parsing.rs
Comment (f4fez) [reply src/parsing.rs]: In 7d0bad1: the fixed-position fields already use named offset consts (`TIMESTAMP_OFFSET`/`SIGNATURE_OFFSET`/`FLAGS_OFFSET`/`MIN_LEN`). The location/feature blocks are optional and flag-gated, so their offsets cannot be fixed consts — I made that relative arithmetic panic-safe with `checked_add` instead.

**Comment by andrewdavidmackenzie** [comment]
I think that's all the code review comments, and patch coverage is much better.

Let me know when you consider it done and ready for review and I'll merge.

**Comment by f4fez** [comment]
Ok great.
You can merge.

**Comment by andrewdavidmackenzie** [comment]
Tested on my mac connecting to meshcore heltec V3.
I must look into connecting it also by serial/USB in the future - suggestions welcome.

Let me know when you can confirm master is OK, and then I can put in some work on release automation and do a release.
